keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,740 @@
+    No
+
+###### Does enabling the feature change any default behavior?
+
+Yes. Users without requisite privileges cannot alter the volume mode of `VolumeSnapshot`
+when it is being used to create a `PVC`. Users with privileges need to add an 
+annotation to the corresponding `VolumeSnapshotContent` instance if they
+require the volume mode to be converted.
+The default behavior does not make any validations prior to provisioning a volume.
+The volume mode can be converted by any user when a `PVC` is created from a 
+`VolumeSnapshot`.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. Disabling the feature is supported and will fall back to the existing behavior.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The new behaviour will be re enabled. `VolumeSnapshots` created when the feature
+was disabled will not have the new capabilities.
+
+###### Are there any tests for feature enablement/disablement?
+
+We will add unit tests with and without the feature flag enabled. The expectation
+is for new fields in `VolumeSnapshotContent` to be dropped when the feature flag
+is disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+###### How can someone using this feature know that it is working for their instance?
```

Comment:
At what stage will the PVC fail?  This may be a challenge to give good feedback when you move to beta. I suggest thinking about this for your alpha implementation.

Reply:
If the new feature flag is enabled, external-provisioner will be checking whether `allowVolumeModeChange` is set to true if the volume mode is changed when creating a PVC from a VolumeSnapshot. If `allowVolumeModeChange` is set to false, CreateVolume request won't be sent to the CSI driver and a PV won't be created. An event should be reported on PVC indicating "ProvisioningFailed".  User can also check external-provisioner logs to find out why the volume is not created.

Reply:
> User can also check external-provisioner logs to find out why the volume is not created.

Pod creating users cannot see this log.

Reply:
That's true.  They can check events on PVC.  We'll need to think more about the troubleshooting part.